esbuild bundling support external dependencies

<!--
If your issue relates to Lambda it _may_ be better handled in this repository:
  https://github.com/DataDog/datadog-lambda-js

If your issue contains sensitive information consider opening a helpdesk ticket instead:
  https://www.datadoghq.com/support/
-->

When bundling code with esbuild and using the dd-trace esbuild plugin, I run into an issue when resolving all dependencies, as some are specified as [external](https://esbuild.github.io/api/#external) for some useless dependencies I'm not using. Yet, dd-trace plugin still tries to resolve them.

**Expected behaviour**
<!-- What should be happening. -->
external packages should not be resolved by the plugin

**Actual behaviour**
<!-- What is be happening. -->
External packages are still resolved, causing issues when they are not added as dependencies

```
✘ [ERROR] Cannot find module 'tedious/package.json'
Require stack:
- /Users/**/node_modules/dd-trace/packages/datadog-esbuild/index.js
- /Users/**/node_modules/dd-trace/esbuild.js
- /Users/**/xbuild.config.js
- /Users/**/node_modules/@myscope/xbuild/dist/xbuild.js [plugin datadog-esbuild]

    node_modules/knex/lib/dialects/mssql/index.js:89:24:
      89 │     const tds = require('tedious');
         ╵                         ~~~~~~~~~

  This error came from the "onResolve" callback registered here:

    node_modules/dd-trace/packages/datadog-esbuild/index.js:38:8:
      38 │   build.onResolve({ filter: /.*/ }, args => {
```

**Steps to reproduce**
<!-- How can we reproduce this issue in order to diagnose it? Code snippets and sample apps are encouraged! -->

Sample project, with knex for example that has multiple external dependencies not useful, for example mysql if I'm using postgres, etc...

If I still try to add those dependencies, I run into another issue, which is that `mysql2` is declared as instrumented, but the resolve is calling the package.json, which is not included in `mysql2`, causing the export to fail similarly. This would indicate that the list `datadog-instrumentations/src/helpers/hooks.js` is not entirely compatible.

```
✘ [ERROR] Package subpath './package.json' is not defined by "exports" in /Users/**/node_modules/mysql2/package.json [plugin datadog-esbuild]

    node_modules/dd-trace/packages/datadog-esbuild/index.js:45:40:
      45 │       const pathToPackageJson = require.resolve(`${packageName}/package.json`, { paths: [ args.resolveDir ] })
```

I think an easy fix would be to either: 
 - filter resolved modules on being or not listed as external in esbuild build config.
 - add the ability for the user to specify a list of excluded modules when specifying the plugin, that would be complementary to the `instrumented` list provided by datadog.

**Environment**

* **Operation system:** Macos
* **Node.js version:** 16.19
* **Tracer version:** 4.0
* **Agent version:**
* **Relevant library versions:**

<!-- Please provide any parts of your package.json that you're able to share, if possible. -->

--- a/LICENSE-3rdparty.csv
+++ b/LICENSE-3rdparty.csv
@@ -51,6 +51,7 @@ dev,glob,ISC,Copyright Isaac Z. Schlueter and Contributors
 dev,graphql,MIT,Copyright 2015 Facebook Inc.
 dev,int64-buffer,MIT,Copyright 2015-2016 Yusuke Kawasaki
 dev,jszip,MIT,Copyright 2015-2016 Stuart Knightley and contributors
+dev,knex,MIT,Copyright (c) 2013-present Tim Griesser
 dev,mkdirp,MIT,Copyright 2010 James Halliday
 dev,mocha,MIT,Copyright 2011-2018 JS Foundation and contributors https://js.foundation
 dev,multer,MIT,Copyright 2014 Hage Yaapa

--- a/integration-tests/esbuild/basic-test.js
+++ b/integration-tests/esbuild/basic-test.js
@@ -11,6 +11,7 @@ const tracer = require('../../').init() // dd-trace
 const assert = require('assert')
 const express = require('express')
 const http = require('http')
+require('knex') // has dead code paths for multiple instrumented packages
 
 const app = express()
 const PORT = 31415

--- a/integration-tests/esbuild/build.js
+++ b/integration-tests/esbuild/build.js
@@ -9,7 +9,18 @@ esbuild.build({
   outfile: 'out.js',
   plugins: [ddPlugin],
   platform: 'node',
-  target: ['node16']
+  target: ['node16'],
+  external: [
+    // dead code paths introduced by knex
+    'pg',
+    'mysql2',
+    'better-sqlite3',
+    'sqlite3',
+    'mysql',
+    'oracledb',
+    'pg-query-stream',
+    'tedious'
+  ]
 }).catch((err) => {
   console.error(err) // eslint-disable-line no-console
   process.exit(1)

--- a/integration-tests/esbuild/package.json
+++ b/integration-tests/esbuild/package.json
@@ -19,6 +19,7 @@
   "license": "ISC",
   "dependencies": {
     "esbuild": "0.16.12",
-    "express": "^4.16.2"
+    "express": "^4.16.2",
+    "knex": "^2.4.2"
   }
 }

--- a/package.json
+++ b/package.json
@@ -120,6 +120,7 @@
     "graphql": "0.13.2",
     "int64-buffer": "^0.1.9",
     "jszip": "^3.5.0",
+    "knex": "^2.4.2",
     "mkdirp": "^0.5.1",
     "mocha": "8",
     "msgpack-lite": "^0.1.26",

--- a/packages/datadog-esbuild/index.js
+++ b/packages/datadog-esbuild/index.js
@@ -39,7 +39,19 @@ module.exports.setup = function (build) {
 
     if (args.namespace === 'file' && packagesOfInterest.has(packageName)) {
       // The file namespace is used when requiring files from disk in userland
-      const pathToPackageJson = require.resolve(`${packageName}/package.json`, { paths: [ args.resolveDir ] })
+
+      let pathToPackageJson
+      try {
+        pathToPackageJson = require.resolve(`${packageName}/package.json`, { paths: [ args.resolveDir ] })
+      } catch (err) {
+        if (err.code === 'MODULE_NOT_FOUND') {
+          console.warn(`Unable to open "${packageName}/package.json". Is the "${packageName}" package dead code?`)
+          return
+        } else {
+          throw err
+        }
+      }
+
       const pkg = require(pathToPackageJson)
 
       if (DEBUG) {

